Sources/SwiftSyntax/SyntaxComparison.swift

```diff
@@ -0,0 +1,80 @@
+//===--------- SyntaxComparison.swift - Syntax Comparison -----------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Utility APIs to compare two difference syntax trees for equivalence.
+//
+//===----------------------------------------------------------------------===//
+
+public enum DifferenceReason {
+  case nodeType, presence, missingNode, additionalNode, trivia, tokenKind, tokenText
+}
+
+public struct TreeDifference {
+  public let node: Syntax
+  public let comparison: Syntax
+  public let reason: DifferenceReason
+
+  public init(lhs: Syntax, rhs: Syntax, reason: DifferenceReason) {
+    self.node = lhs
+    self.comparison = rhs
+    self.reason = reason
+  }
+}
+
+public extension Syntax {
+  func isEquivalent(to other: Syntax, includeTrivia: Bool = false) -> DifferenceReason? {
```

Comment:
Since this returns a `DifferenceReason`, I think `isDifferent` would be a better name. With the current name, I would expect it to return something `true`-ish (i.e. non-`nil`) if the trees are equivalent.

Also, I think it’s worth mentioning that this does not compare children in a doc comment and because of this reason make it private, if possible.

Reply:
Good point 👍